feat: add territory wise sales report

0Pranav opened the pull request:
- add `Expected Closing Date` to Opportunity DocType
- add territory-wise sales report (screenshot below)

<img width="1280" alt="Screenshot 2020-01-27 at 11 04 06 AM" src="https://user-images.githubusercontent.com/6195660/73152207-dd604e80-40c6-11ea-931f-8b8c821fa578.png">

Commit:
enhancement: add territory wise sales report

deepeshgarg007 commented on erpnext/crm/doctype/social_media_account/social_media_account.json:
```diff
@@ -0,0 +1,89 @@
+   "print": 1,
+   "read": 1,
+   "report": 1,
+   "role": "System Manager",
```
Add  permission for Sales Manager as well

deepeshgarg007 commented on erpnext/crm/doctype/social_media_account/twitter.py:
```diff
@@ -0,0 +1,31 @@
+import urllib
+
+
+def get_request_token_url(consumer_key,consumer_key_secret, callback):
```
I think this file is pushed by mistake. Please remove

deepeshgarg007 commented on erpnext/selling/report/territory_wise_sales/territory_wise_sales.js:
```diff
@@ -0,0 +1,14 @@
+// For license information, please see license.txt
+/* eslint-disable */
+
+frappe.query_reports["Territory wise Sales"] = {
```
Report name should be title case

deepeshgarg007 commented on erpnext/selling/report/territory_wise_sales/territory_wise_sales.js:
```diff
@@ -0,0 +1,14 @@
+/* eslint-disable */
+
+frappe.query_reports["Territory wise Sales"] = {
+	"filters": [
```
Add company filter as well

deepeshgarg007 commented on erpnext/selling/report/territory_wise_sales/territory_wise_sales.js:
```diff
@@ -0,0 +1,14 @@
+		{
+			fieldname:"expected_closing_date",
+			label: __("Expected Closing Date"),
+			fieldtype: "DateRange",
```
Make this filter mandatory

deepeshgarg007 commented on erpnext/selling/report/territory_wise_sales/territory_wise_sales.py:
```diff
@@ -0,0 +1,137 @@
+	conditions = ""
+
+	if filters.from_date and filters.to_date:
+		conditions = " WHERE expected_closing between %(from_date)s and %(to_date)s"
```
transaction date should be used here instead of expected closing date

deepeshgarg007 commented on erpnext/selling/report/territory_wise_sales/territory_wise_sales.py:
```diff
@@ -0,0 +1,137 @@
+	sales_invoices = get_sales_invoice(sales_orders)
+
+	for territory in frappe.get_all("Territory"):
+		territory_opportunities = list(filter(lambda x: x.territory == territory.name, opportunities)) if opportunities and opportunities else None
```
```suggestion
		territory_opportunities = list(filter(lambda x: x.territory == territory.name, opportunities)) if opportunities else []
```

deepeshgarg007 commented on erpnext/selling/report/territory_wise_sales/territory_wise_sales.py:
```diff
@@ -0,0 +1,137 @@
+
+	for territory in frappe.get_all("Territory"):
+		territory_opportunities = list(filter(lambda x: x.territory == territory.name, opportunities)) if opportunities and opportunities else None
+		t_opportunity_names = [t.name for t in territory_opportunities] if territory_opportunities else None
```
```suggestion
		t_opportunity_names = [t.name for t in territory_opportunities] if territory_opportunities else []
```

deepeshgarg007 replied:
Do the same for other condition checks as well

deepeshgarg007 commented on erpnext/crm/doctype/social_media_account/social_media_account.js:
```diff
@@ -0,0 +1,8 @@
+// Copyright (c) 2020, Frappe Technologies Pvt. Ltd. and contributors
```
Remove this file and other related as files as well

deepeshgarg007 commented on erpnext/selling/report/territory_wise_sales/territory_wise_sales.py:
```diff
@@ -0,0 +1,137 @@
+
+	total = 0
+	for doc in doclist:
+		total = total + doc.get(amount_field, 0)
```
```suggestion
		total += doc.get(amount_field, 0)
```

deepeshgarg007 reviewed (changes requested):
Add Report link in CRM module view as well

Commit:
various fixes from review
- Rename the report to title case
- Added company filter
- fixed checks in filter and += operator for total

Commit:
fix: return empty array

deepeshgarg007 commented on erpnext/selling/report/territory_wise_sales/territory_wise_sales.py:
```diff
@@ -0,0 +1,145 @@
+	sales_invoices = get_sales_invoice(sales_orders)
+
+	for territory in frappe.get_all("Territory"):
+		territory_opportunities = list(filter(lambda x: x.territory == territory.name, opportunities)) if opportunities and opportunities else []
```
territory_opportunities = []
if opportunities:
         territory_opportunities = list(filter(lambda x: x.territory == territory.name, opportunities))
         t_opportunity_names = [t.name for t in territory_opportunities] if territory_opportunities else []

Commit:
fix: better syntax for checking empty arrays

Commit:
fix: filters after rename

deepeshgarg007 committed:
fix: Add report link in module view and fix date filter

deepeshgarg007 committed:
fix: Remove debug param
